### ISSUE TITLE
Improve documentation and facilitate usage with Mindboggle101

### ISSUE THREAD

**Issue by tvercaut** [cai4cai/gif-python#1]
From @sirisaacnewton540 's experience, it is not straightforward to use this repo to get DKT parcellations using the Mindboggle101 dataset as source multi-atlas.

This should be the main usage of this repo and should therefore be clearly documented, providing all required config files to run this easily.

**Comment by tvercaut** [comment]
This is probably the expected csv file for mindboggle:
[mindboggle_atlases_structures_info.csv](https://github.com/user-attachments/files/24133049/mindboggle_atlases_structures_info.csv)

**Comment by LorenaGarcia-Foncillas** [comment]
There are 6 labels present in the labels.DKT31.manual+aseg.nii.gz files that are not included in the structures_info.csv spreadsheet:

Missing Labels
- 1001 (ctx-lh-bankssts): Present in 2 cases
    - Cases: MMRR-21-6, OASIS-TRT-20-16
- 1032 (ctx-lh-frontalpole): Present in 3 cases
    - Cases: HLN-12-12, MMRR-21-1, OASIS-TRT-20-10
- 1033 (ctx-lh-temporalpole): Present in 9 cases
    - Cases: HLN-12-2, MMRR-21-1, MMRR-21-19, NKI-RS-22-10, NKI-RS-22-12, NKI-RS-22-13, NKI-RS-22-22, OASIS-TRT-20-2, OASIS-TRT-20-4
- 2001 (ctx-rh-bankssts): Present in 1 case
    - Cases: MMRR-21-10
- 2032 (ctx-rh-frontalpole): Present in 7 cases
    - Cases: HLN-12-3, HLN-12-5, HLN-12-9, MMRR-21-17, MMRR-21-9, OASIS-TRT-20-12, OASIS-TRT-20-20
- 2033 (ctx-rh-temporalpole): Present in 10 cases
    - Cases: HLN-12-2, HLN-12-8, MMRR-21-20, NKI-RS-22-12, NKI-RS-22-18, NKI-RS-22-9, OASIS-TRT-20-12, OASIS-TRT-20-14, OASIS-TRT-20-3, OASIS-TRT-20-6

These labels appear to be part of the standard FreeSurfer parcellation scheme (see https://surfer.nmr.mgh.harvard.edu/fswiki/FsTutorial/AnatomicalROI/FreeSurferColorLUT):

  - bankssts = Banks of the Superior Temporal Sulcus
  - frontalpole = Frontal Pole
  - temporalpole = Temporal Pole

Since all these structures use the ctx- prefix (cortex), they should be classified as Grey Matter and assigned tissue type [2].


> [!IMPORTANT]
>  Add the following entries to structures_info.csv:
>  -   1001,ctx-lh-bankssts,[2]
>  -   1032,ctx-lh-frontalpole,[2]
>  -   1033,ctx-lh-temporalpole,[2]
>  -   2001,ctx-rh-bankssts,[2]
>  -   2032,ctx-rh-frontalpole,[2]
>  -   2033,ctx-rh-temporalpole,[2]

**Comment by tvercaut** [comment]
What about those? Aaron previously mentioned them together with the 6 labels you list here:
```
label_to_name[72] = "5th ventricle"
label_to_name[80] = "non-WM-hypointensities"
```

Here is a picture showing label presence in the mindboggle101 dataset:

<img width="2609" height="2768" alt="Image" src="https://github.com/user-attachments/assets/b4d727b3-f000-496b-850b-9ce1212f10a9" />

**Comment by tvercaut** [comment]
Attached are the results from [Aaron's MICCAI paper](https://arxiv.org/abs/2404.10572). I now remember he discarded labels [1001, 1032, 1033, 2001, 2032, 2033]. I think he also suggested afterward to discard labels [72, 80].  I'll try and track this down.

[supplementary_material_page2_rotated.pdf](https://github.com/user-attachments/files/24187179/supplementary_material_page2_rotated.pdf)

**Comment by tvercaut** [comment]
Found this code snippet that I think was used to remove the sparsest 6 labels from the "GT" to get a cleaner version (replacing by neighbour label values). It would make sense to provide this here in this repo but also archive a version of the cleaned version of Mindboggle we end up using.

```python
import nibabel as nib
import os
import numpy as np
from glob import glob

labels_to_replace = [1001, 1032, 1033, 2001, 2032, 2033]

p_in_all = glob("./input_data/*/labels.nii.gz")
p_out_all = [p.replace(".nii.gz", "_cleaned.nii.gz") for p in p_in_all]

def get_neighbors(coords):
    x, y, z = coords
    
    # Define offsets for all possible neighboring coordinates
    offsets = np.array([
        [-1, -1, -1], [-1, -1, 0], [-1, -1, 1],
        [-1, 0, -1], [-1, 0, 0], [-1, 0, 1],
        [-1, 1, -1], [-1, 1, 0], [-1, 1, 1],
        [0, -1, -1], [0, -1, 0], [0, -1, 1],
        [0, 0, -1], [0, 0, 1],
        [0, 1, -1], [0, 1, 0], [0, 1, 1],
        [1, -1, -1], [1, -1, 0], [1, -1, 1],
        [1, 0, -1], [1, 0, 0], [1, 0, 1],
        [1, 1, -1], [1, 1, 0], [1, 1, 1]
    ])
    
    # Add the offsets to the input coordinates
    neighbor_coords = coords + offsets
    
    return neighbor_coords

for p_in, p_out in list(zip(p_in_all, p_out_all)):
    print(p_in)
    nii = nib.load(p_in)
    data = nii.get_fdata()

    data_out = np.copy(data)
    
    for l_in in labels_to_replace:

        locs = np.where(data==l_in)

        for i,j,k in zip(*locs):
            neighbor_vals = []
            #print("voxel: ", i,j,k)
            neighbors = get_neighbors([i,j,k])
            #print(neighbors)

            for n in neighbors:
                neighbor_val = data[n[0], n[1], n[2]]
                if neighbor_val != l_in:
                    neighbor_vals.append(neighbor_val)

            #print("neighbor label counts: ")
            if len(neighbor_vals) == 0: raise Exception("no neighbors found...")
            
            labels, counts = np.unique(np.array(neighbor_vals), return_counts=True)
            #print(labels, counts)

            max_counts_idx = np.argmax(counts)

            max_label = labels[max_counts_idx]

            assert(max_label not in labels_to_replace), f"{max_label=} in {labels_to_replace=}"

            #print("max_label", max_label)

            print(f"change {l_in} to {max_label}")
            data_out[i, j, k] = max_label

        
    print("number_of_changed_voxels", np.sum(data_out != data))

    nii_out = nib.Nifti1Image(dataobj=data_out, affine=nii.affine)
    nib.save(nii_out, filename=p_out)
```

**Comment by LorenaGarcia-Foncillas** [comment]
Thanks for sharing the code to remove sparse labels! 
Should we then remove all 8 labels [1001, 1032, 1033, 2001, 2032, 2033, 72, 80] from the atlas data using the script and replace them with neighbouring voxel values (creating a "cleaned" version)?

**Comment by tvercaut** [comment]
Some further notes on "problematic" labels.

### 5th Ventricle (Label 72, Average DSC = 0.0%)
According to [freesurfer.net CMA segmentation protocol](https://freesurfer.net/fswiki/SubcorticalSegmentation):
*“The 5th ventricle, referred to as the cavum septum pellucidum, is a bit of an enigma that tends to only occur on occasion. It’s not reported in any manuscripts, but it was confirmed by neuroradiologist P. Ellen Grant.”*

It is present in only 35 of the 59 Mindboggle cases that were part of the multi-atlas. If present, the label occupies no more than a few voxels. Hence, the training data did not contain sufficient examples for the CNN to predict this structure. **We suggest discarding this label in the future.**

### Non-WM-Hypointensities (Label 80, Average DSC = 0.0%) and WM-Hypointensities (Label 77, Average DSC = 53.0%)
These are areas of reduced intensity in gray matter and white matter, respectively. Due to their variable location, these labels are not suitable for multi-atlas propagation, resulting in low DSCs. **These labels should be replaced with anatomical labels in the Mindboggle ground truth.**

### ctx-rh-unknown (Label 1000, Average DSC = 66.2%) and ctx-lh-unknown (Label 2000, Average DSC = 68.7%)
These labels seem to be an artifact of merging subcortical FreeSurfer labels and manual cortical labels in the Mindboggle dataset. They appear at the border between gray and white matter, for example surrounding the cingulate gyrus (cingulum). **These labels should be removed or corrected manually in the Mindboggle ground truth.**

### Optic Chiasm (Label 85, Average DSC = 66.8%)
The optic chiasm is a small structure (average volume of 211 mm³). When comparing the Mindboggle ground truth to manual annotations performed internally at KCL, it was noticed that the optic chiasm was not always correctly labeled in the Mindboggle ground truth. **These labels should be corrected manually in the Mindboggle ground truth.**

### Ventricles and Choroid Plexus
- Right inferior lateral ventricle (Label 44, Average DSC = 75.6%)  
- Left inferior lateral ventricle (Label 5, Average DSC = 75.6%)  
- Right choroid plexus (Label 63, Average DSC = 78.1%)  
- Left choroid plexus (Label 31, Average DSC = 78.3%)  

Although acceptable DSCs are achieved for these structures, their small volume resulted in a higher number of outliers and smaller DSCs compared to other structures. Furthermore, the choroid plexus is located within the ventricle and exhibits low contrast on T1-weighted MRI, which further reduces segmentation accuracy. **Depending on the use case, accuracy for these structures may be sufficient to keep.**

### CSF (Label 24, Average DSC = 78.4%)
This label does not refer to all cerebrospinal fluid surrounding the brain (this is included in the background label), but only to the Transverse Cerebral Fissure (TCF). As described in the [CMA segmentation protocol](https://freesurfer.net/fswiki/CMA?action=AttachFile&do=view&target=CMA_segmentation_may_2004.pdf):  
*“The TCF lies outside of the brain exterior and is filled with extraventricular (subarachnoidal) CSF. In some ways, it is an imprecise label because in its posterior extent, what is extracted and labeled as CSF will include TCF and the pineal gland. Though this label is imprecise, it is necessary as it ensures we do not include any TCF in the third ventricle.”*  

This imprecision and the lack of contrast with adjacent CSF result in reduced DSC. **This label may be kept since its purpose is mainly to guide the annotation of the 3rd ventricle.**

**Comment by tvercaut** [comment]
> Should we then remove all 8 labels [1001, 1032, 1033, 2001, 2032, 2033, 72, 80] from the atlas data using the script and replace them with neighbouring voxel values (creating a "cleaned" version)?

I think it makes sense to list all labels in `mindboggle_atlases_structures_info.csv` as this shouldn't cause any issue even if we "clean-up" problematic labels.

The script above will only work to replace very small labels. If it works for [1001, 1032, 1033, 2001, 2032, 2033, 72, 80, 77], let's do that, otherwise, as an initial step, let's generate a cleaned version of mindboggle101 by removing only [1001, 1032, 1033, 2001, 2032, 2033]. We can then discuss how do best handle [72, 80, 77, 1000, 2000, 85].

@ivorymarina had some views on the Optic Chiasm (Label 85). She also referred to the CerebrA paper:
Manera, A. L., Dadar, M., Fonov, V., & Collins, D. L. (2020). CerebrA, registration and manual label correction of Mindboggle-101 atlas for MNI-ICBM152 template. Scientific Data, 7(1), 237.
https://doi.org/10.1038/s41597-020-0557-9

It might make sense to perform a more advance "clean-up" using registration of unique cases to CerebrA but I think this is work that @sirisaacnewton540 could look into as a more longer term question.

**Comment by tvercaut** [comment]
As a way of facilitating usage of Mindboggle, we should also provide an "ITK-SnAP Label Description File". Here is one I scavenged, it would make sense to also add "problematic" labels to it.

[itksnap_label_descriptions_mindboggle101.txt](https://github.com/user-attachments/files/24188289/itksnap_label_descriptions_mindboggle101.txt)

**Comment by ivorymarina** [comment]
Thanks Tom,

I might need a bit of context on this.
Will see you at the office today if you are joining my PhD meeting at 3.30 with Jonathan.

Kind regards,
Marina.

On 16 Dec 2025, at 10:07, Tom Vercauteren ***@***.***> wrote:

﻿
[https://avatars.githubusercontent.com/u/1614505?s=20&v=4]tvercaut left a comment (cai4cai/gif-python#1)<https://github.com/cai4cai/gif-python/issues/1#issuecomment-3659758096>

As a way of facilitating usage of Mindboggle, we should also provide an "ITK-SnAP Label Description File". Here is one I scavenged, it would make sense to also add "problematic" labels to it.

itksnap_label_descriptions_mindboggle101.txt<https://github.com/user-attachments/files/24188289/itksnap_label_descriptions_mindboggle101.txt>

**Comment by LorenaGarcia-Foncillas** [comment]
**Changes in Latest Commit ([0f314ce](https://github.com/cai4cai/gif-python/commit/0f314ce2c614b025c24b2c0f95a3543d41eb38e9))**

1. Added INSTALL.md: Complete installation guide including:
    - Software requirements (Python, NiftyReg, NiftySeg)
    - Data folder structure with expected file organization
    - Instructions for obtaining Mindboggle101 data from Synapse
    - Details on structures_info.csv and tissues_info.csv requirements
    - Note about the 6 sparse labels and the clean_small_labels.py script
2. Added run_segmentation_basic.py: Simplified script for running multi-atlas segmentation on a single subject
3. Added clean_small_labels.py: Script to remove sparse labels by replacing them with neighbouring voxel values (@tvercaut approach shared in this thread)

**Comment by p-singh-kcl** [comment]
I've identified and am working on the registration failures affecting approximately 50% of my dataset. The issue was suboptimal default parameters for my specific data characteristics, rather than a fundamental algorithmic failure.

The failures manifested as:
**[NiftyReg WARNING] The current level reached the maximum number of iteration
[NiftyReg WARNING] Function: reg_f3d<T>::ComputeJacobianBasedPenaltyTerm()
[NiftyReg WARNING] The folding correction scheme failed**

Initial investigation ruled out data heterogeneity (contrast enhancement, acquisition protocol variations, pathology). All failing cases were standard pre-contrast T1-weighted clinical scans with adequate brain extraction.

The default LNCC (Local Normalized Cross-Correlation) kernel size and associated optimization parameters were insufficient for convergence on scans with certain intensity characteristics. By systematically adjusting the LNCC kernel size and related regularization terms, I achieved a bit better registration on previously failing cases (not perfect).

I'm currently conducting a parameter sweep to identify optimal settings that maximize registration success while maintaining anatomical plausibility (minimizing Jacobian determinant violations). Preliminary results indicate that the LNCC kernel size is a critical factor, likely due to its influence on the local intensity neighborhood used for similarity computation.

**Comment by p-singh-kcl** [comment]
While setting up the pipeline with different atlases, several of us noticed that manually creating the **structure_info.csv** file is both time-consuming and error-prone. The process requires extracting all unique labels from the atlas, matching them to FreeSurferColorLUT names, formatting everything correctly, and ensuring there are no typos or inconsistencies that can break the pipeline.

This becomes an especially significant bottleneck when working with non-standard atlases, where the same manual steps must be repeated each time. To address this, I wrote a script that automates the entire process. It reads the atlas segmentation file, extracts label indices, maps them to the appropriate FreeSurfer naming conventions, and generates a ready-to-use **structure_info.csv** and **tissue_info.csv**. It also handles common edge cases such as excluding background labels and managing bilateral structures.

[lut_from_pdf.csv](https://github.com/user-attachments/files/25162938/lut_from_pdf.csv)
and
[generate_gif_structures.ipynb](https://github.com/user-attachments/files/25162952/generate_gif_structures.ipynb)
or 
[generate_gif_structures.py](https://github.com/user-attachments/files/25162937/generate_gif_structures.py)


I think this could be a valuable addition to the pipeline as an optional utility, reducing setup friction and making it easier for users to work with diverse atlases. It could live as a standalone preprocessing script, an optional pipeline flag, or a helper tool within the repository.

Let me know if you think this would be worth including.
